server/snapshot.go

```diff
@@ -325,43 +360,60 @@ func (r *raftFSM) restoreClientsFromSnapshot(serverSnap *spb.RaftSnapshot) error
 	return nil
 }
 
-func (r *raftFSM) restoreChannelsFromSnapshot(serverSnap *spb.RaftSnapshot, inNewRaftCall bool) error {
+func (r *raftFSM) restoreChannelsFromSnapshot(serverSnap *spb.RaftSnapshot, inNewRaftCall bool) (bool, error) {
 	s := r.server
 
+	shouldSnapshot := false
+
 	var channelsBeforeRestore map[string]*channel
 	if !inNewRaftCall {
 		channelsBeforeRestore = s.channels.getAll()
 	}
 	for _, sc := range serverSnap.Channels {
 		c, err := s.lookupOrCreateChannel(sc.Channel)
 		if err != nil {
-			return err
+			return false, err
 		}
 		// Keep track of first/last sequence in this channel's snapshot.
 		// It can help to detect that all messages in the leader have expired
 		// and this is what we should report/use as our store first/last.
-		// This is running under the FSM lock which protects this field.
 		atomic.StoreUint64(&c.firstSeq, sc.First)
 
 		// Even on startup (when inNewRaftCall==true), we need to call
 		// restoreMsgsFromSnapshot() to make sure our store is consistent.
-		// If we skip it (like we use to), then it is possible that this
+		// If we skip it (like we used to), then it is possible that this
 		// node misses some messages from the snapshot and will then start
 		// to apply remaining logs. It could even then become leader in
 		// the process with missing messages. So if we need to restore some
 		// messages and fail, keep trying until a leader is avail and
 		// able to serve this node.
-		for {
+		ok := false
+		for i := 0; i < restoreMsgsAttempts; i++ {
 			if err := r.restoreMsgsFromSnapshot(c, sc.First, sc.Last, false); err != nil {
-				s.log.Errorf("unable to restore messages, can't start until leader is available")
-				time.Sleep(time.Second)
+				s.log.Errorf("channel %q - unable to restore messages, can't start until leader is available",
+					sc.Channel)
+				time.Sleep(restoreMsgsSleepBetweenAttempts)
 			} else {
+				ok = true
 				break
 			}
 		}
+		if !ok {
+			err = fmt.Errorf("channel %q - unable to restore messages, aborting", sc.Channel)
+			s.log.Fatalf(err.Error())
+			// In tests, we use a "dummy" logger, so process will not exit
+			// (and we would not want that anyway), so make sure we return
+			// an error.
+			return false, err
```

Comment:
Can we even get here with Fatalf?

Reply:
Read comment above ;-). Yes, we can and do in all tests since our test logger does not exit. It would be bad if we did because all tests that trigger a fatalf would cause the test suite to stop (due to call to os.Exit()).